Sources/MongoSwift/MongoCollection+Indexes.swift

```diff
@@ -335,20 +342,28 @@ extension MongoCollection {
      * - Parameters:
      *   - session: Optional `ClientSession` to use when executing this command
      *
+     * - Returns: A `MongoCursor` over the `IndexModel`s.
+     *
+     * - Throws: `UserError.logicError` if the provided session is inactive.
+     */
+    public func listIndexes(session: ClientSession? = nil) throws -> MongoCursor<IndexModel> {
+        let operation = ListIndexesOperation(collection: self)
+        return try self._client.executeOperation(operation, session: session)
+    }
+
+    /**
+     * Retrieves a list of names of the indexes currently on this collection.
+     *
+     * - Parameters:
+     *   - session: Optional `ClientSession` to use when executing this command
+     *
      * - Returns: A `MongoCursor` over the index names.
      *
      * - Throws: `UserError.logicError` if the provided session is inactive.
      */
-    public func listIndexes(session: ClientSession? = nil) throws -> MongoCursor<Document> {
-        let opts = try encodeOptions(options: Document(), session: session)
-
-        return try MongoCursor(client: self._client, decoder: self.decoder, session: session) { conn in
-            self.withMongocCollection(from: conn) { collPtr in
-                guard let cursor = mongoc_collection_find_indexes_with_opts(collPtr, opts?._bson) else {
-                    fatalError(failedToRetrieveCursorMessage)
-                }
-                return cursor
-            }
-        }
+    public func listIndexNames(session: ClientSession? = nil) throws -> [String] {
+        let operation = ListIndexesOperation(collection: self)
+        let models = try self._client.executeOperation(operation, session: session)
+        return models.map { $0.options?.name ?? $0.defaultName }
```

Comment:
the server should always return a name field, so lets throw an error here instead of supplying a name on the fly. If a name doesn't show up here, it's likely to due a bug in the driver somewhere, and we don't want the behavior caused by that bug to be hidden.

An interesting concept in swift that is related to this is `rethrows`: Similar to how you can mark a function as `throws`, meaning it may return errors and must be invoked with `try`, you can mark a method as `rethrows`, meaning it will only be `throws` if a closure passed to it is `throws`.

The `map` method on `Array` is `rethrows`, so we can still use it here with a `throws` closure:
```swift
return try models.map { model in 
    guard let name = model.options?.name else {
        throw RuntimeError.internalError("Server response missing a 'name' field")
    }
    return name
}
```